### PR TITLE
Refactored job event waiting to use a new JobWatcher abstraction

### LINKED ISSUE #1475: Refactor waiting for job events
There are a couple of places where we currently wait for a number of `up` and / or `down` job events, and the code is pretty similar:

* [`cli/scale.go`](https://github.com/flynn/flynn/blob/master/cli/scale.go#L120-L161)
* [`test/test_deployer.go`](https://github.com/flynn/flynn/blob/master/test/test_scheduler.go#L43-L73)

We will also need to do this same thing to close #1260 (see #1474), so I think it would be good to refactor to something like:

```go
watcher, err := client.WatchJobEvents(app.ID)
if err != nil {
	return err
}
defer watcher.Close()

// make some job changes

err := watcher.WaitFor(controller.JobEvents{
	"web":    {"up": 1},
	"worker": {"down": 2},
})
```

Both examples of the code print some information when a job event is received, so we could provide a callback to be fired on each event, or give the watcher a logger, or something else.

### PR DESCRIPTION
**Pull request by jzila**
Fixes #1475

Implemented an abstraction for waiting for job events instead of
re-implementing the channel loop each time we need to watch for jobs.

In instances where we need to multiplex job streams with other channels,
the original code is still in place.

### CODE REVIEW

**Comment by jzila** [comment]
The last 2 CI failures have been intermittent failures independent of this change.

File: controller/client/client.go [hunk 1]
```diff
@@ -351,6 +435,38 @@ func (c *Client) StreamJobEvents(appID string, output chan *ct.JobEvent) (stream
...
+func (c *Client) WatchJobEvents(appID string) (*JobWatcher, error) {
+	events := make(chan *ct.JobEvent)
+
+	stream, err := c.StreamJobEvents(appID, events)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return newJobWatcher(events, stream), nil
```
Comment (lmars) [inline controller/client/client.go]: I don't think there is need for the blank lines in this method

File: controller/client/client.go [hunk 2]
```diff
@@ -351,6 +435,38 @@ func (c *Client) StreamJobEvents(appID string, output chan *ct.JobEvent) (stream
...
+}
+
+func (c *Client) ExpectedScalingEvents(actual, expected map[string]int, clusterSize int) ct.JobEvents {
+	events := make(ct.JobEvents)
```
Comment (lmars) [inline controller/client/client.go]: Add a `len(expected)` capacity hint.

File: controller/client/client.go [hunk 3]
```diff
@@ -351,6 +435,38 @@ func (c *Client) StreamJobEvents(appID string, output chan *ct.JobEvent) (stream
...
+	stream, err := c.StreamJobEvents(appID, events)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return newJobWatcher(events, stream), nil
+}
+
+func (c *Client) ExpectedScalingEvents(actual, expected map[string]int, clusterSize int) ct.JobEvents {
```
Comment (lmars) [inline controller/client/client.go]: Does this need to be a method on the client? Given it just creates `ct.JobEvents`, maybe it should be in the types package?

File: controller/client/client.go [hunk 4]
```diff
@@ -351,6 +435,38 @@ func (c *Client) StreamJobEvents(appID string, output chan *ct.JobEvent) (stream
...
+	events := make(ct.JobEvents)
+	for typ, count := range expected {
+		diff := count
+		val, ok := actual[typ]
+		if ok {
+			diff = count - val
+		}
+		if typ == "omni" {
+			diff *= clusterSize
+		}
```
Comment (lmars) [inline controller/client/client.go]: This check is incorrect, the type being named "omni" does not indicate the type is an omni job (it just so happens that this is the case in the tests), the correct check is against the [process type's Omni field](https://github.com/flynn/flynn/blob/master/controller/types/types.go#L49).

Note this handling of omni jobs will most likely be simplified in the next implementation of the scheduler, see #1484.

File: controller/client/client.go [hunk 5]
```diff
@@ -31,6 +32,89 @@ type Client struct {
 	*httpclient.Client
 }
 
+type JobWatcher struct {
+	events chan *ct.JobEvent
+	stream stream.Stream
+}
```
Comment (lmars) [inline controller/client/client.go]: This should be in the types package.

File: controller/client/client.go [hunk 6]
```diff
@@ -31,6 +32,89 @@ type Client struct {
...
+				err := callback(e)
+				if err != nil {
+					return err
+				}
+			}
+			if jobEventsEqual(expected, actual) {
+				return nil
+			}
+		case <-time.After(timeout):
+			return fmt.Errorf("Timed out waiting for job events. Waited %d seconds.\nexpected: %s\nactual: %s", timeout, toJson(expected), toJson(actual))
```
Comment (lmars) [inline controller/client/client.go]: I think the default printing of the `map` type (i.e. `%v`) rather than converting to JSON should suffice in the error output, e.g.:

```
map[web:map[up:1] worker:map[down:3]]
```

File: test/test_cli.go [hunk 7]
```diff
@@ -189,7 +196,9 @@ func (s *CLISuite) TestPs(t *c.C) {
 	for _, j := range jobs {
 		t.Assert(j, Matches, "echoer")
 	}
-	t.Assert(app.flynn("scale", "echoer=0"), Succeeds)
+	scale := app.flynn("scale", "echoer=0")
+	app.waitFor(ct.JobEvents{"echoer": {"down": 3}})
```
Comment (lmars) [inline test/test_cli.go]: This check is a duplication of the fact that the scale command succeeds

File: test/test_cli.go [hunk 8]
```diff
@@ -231,17 +242,18 @@ func (s *CLISuite) TestScale(t *c.C) {
 
 	// unchanged processes shouldn't appear in output
 	scale = app.flynn("scale", "echoer=3", "printer=0")
+	app.waitFor(ct.JobEvents{"printer": {"down": 2}})
 	t.Assert(scale, Succeeds)
 	t.Assert(scale, SuccessfulOutputContains, "printer: 2=>0")
-	t.Assert(scale, c.Not(SuccessfulOutputContains), "echoer")
+	t.Assert(scale, c.Not(OutputContains), "echoer")
```
Comment (lmars) [inline test/test_cli.go]: Why the change in assertion?

File: test/test_cli.go (hunk 7 above)
Comment (jzila) [inline test/test_cli.go]: Ostensibly yes, but I've had TestScale fail without this check.

File: test/test_cli.go (hunk 8 above)
Comment (jzila) [inline test/test_cli.go]: SuccessfulOutputContains is analogous to `Success AND OutputContains`. This means that `c.Not(SuccessfulOutputContains)` is analogous to `NOT Success OR NOT OutputContains`, when I want to explicitly check that the output doesn't contain a specific string.

File: controller/client/client.go (hunk 5 above)
Comment (jzila) [inline controller/client/client.go]: Then I need to move all the JobWatcher receiving methods to the types package as well. From what I could tell, that package was avoiding having types that receive methods. Did I misunderstand? Is there a cleaner way to do this?

File: test/test_cli.go (hunk 7 above)
Comment (lmars) [inline test/test_cli.go]: Hmm is that an intermittent failure? It would be good to understand the issue, as this check seems redundant.

File: controller/client/client.go (hunk 5 above)
Comment (lmars) [inline controller/client/client.go]: Ok yes I missed that point, fine being here then.

**Comment by jzila** [comment]
@lmars does your latest comment indicate that we should remove the TestScale watcher waits, or are we good to go?

**Comment by lmars** [comment]
> does your latest comment indicate that we should remove the TestScale watcher waits

I am just curious to understand why you needed to change `TestPs` because `TestScale` was failing?

Other than that, please squash to a single commit (e.g. `controller,cli: Refactor waiting for job events`) and we are good to merge.

**Comment by jzila** [comment]
Sorry, I made the same change in TestScale so I was confused what you were asking about. Yes, both `TestScale` and `TestPs` fail intermittently without the check.

**Comment by lmars** [comment]
Thanks!
